fix(freertos): Limit idle task name length copy operation and ensure null-termination of the idle task name string

sudeep-mohanty opened the pull request:
This PR:
    - Limits the idle task name length copy operation to prevent Out-of-bounds memory access warnings from static code analyzers.
    - Fixes a bug where in the idle task name could be non null-terminated string for SMP configuration.

Description
-----------
- In the function `prvCreateIdleTasks()`, we have the operation -
```
       for( xIdleTaskNameIndex = ( BaseType_t ) 0; xIdleTaskNameIndex < ( BaseType_t ) configMAX_TASK_NAME_LEN; xIdleTaskNameIndex++ )
       {
          cIdleName[ xIdleTaskNameIndex ] = configIDLE_TASK_NAME[ xIdleTaskNameIndex ];
  
          /* Don't copy all configMAX_TASK_NAME_LEN if the string is shorter than
           * configMAX_TASK_NAME_LEN characters just in case the memory after the
           * string is not accessible (extremely unlikely). */
          if( cIdleName[ xIdleTaskNameIndex ] == ( char ) 0x00 )
          {
             break;
          }
          else
          {
             mtCOVERAGE_TEST_MARKER();
          }     
       }
  ```

- Static code analyzers could flag this copy operation as an Out-of-bounds memory access because they see `configIDLE_TASK_NAME` as an array of 5 bytes (I, D, L, E, \0) but the loop running for `configMAX_TASK_NAME_LEN` iterations which could be more than 5.
- They also do not assume that a `\0` character is present in the `configIDLE_TASK_NAME` array and hence can not predict that the loop will break before an Out-of-bounds memory access is made.
- Therefore, this PR limits the copy operation to the minimum of either `configIDLE_TASK_NAME` or `configMAX_TASK_NAME_LEN`. This ensures that the copy operation runs for exactly the required number of iterations to copy the idle task name, 5 by default.
- The PR also fixes a bug with null-termination of the `IDLE` task name if `strlen(configIDLE_TASK_NAME) = configMAX_TASK_NAME_LEN - 1` when SMP configuration is enabled. The current code would append the core ID to the task name and in the process overwrites the null terminator. It then exists the loop as there is no more space to add a null terminator. 

Test Steps
-----------
- Run a static code analysis of `tasks.c` on a tool like Coverity.

Checklist:
----------
- [x] I have tested my changes. No regression in existing tests.

Related Issue
-----------
None

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

jasonpcarroll reviewed (commented):
Doing this means you no longer need the logic within the for loop that limits how many characters are copied into the buffer. you might as well remove that logic as well. Also, since you are already using strlen why not just replace all of this code with strncpy? https://cplusplus.com/reference/cstring/strncpy/

sudeep-mohanty commented:
@jasonpcarroll I've removed the `break`-ing logic inside the loop and updated it to use standard string operations.

On a second note, just discovered that there is a potential problem in handling the null-termination of the `IDLE` task name in case of SMP configuration. The current logic would not null-terminate the idle task array if `strlen(configIDLE_TASK_NAME) = configMAX_TASK_NAME_LEN - 1`. The logic would append the `xCoreID`, in the process overwriting the null character and then not adding a null-character any more.

I've updated the logic to use strncat instead.

sudeep-mohanty commented:
> ## [![Quality Gate Failed](https://camo.githubusercontent.com/ce402bcc4cf818eda660128f0b27a891eb8c9f4e0d783303ea667cb0d4b07f95/68747470733a2f2f736f6e6172736f757263652e6769746875622e696f2f736f6e6172636c6f75642d6769746875622d7374617469632d7265736f75726365732f76322f636865636b732f5175616c6974794761746542616467652f71672d6661696c65642d323070782e706e67)](https://sonarcloud.io/dashboard?id=FreeRTOS_FreeRTOS-Kernel&pullRequest=1203) **Quality Gate failed**
> Failed conditions ![](https://camo.githubusercontent.com/bd1d5234b37410841ad8cd9d85c84eeecaec92ae18a95c926ae3a59a17c29a65/68747470733a2f2f736f6e6172736f757263652e6769746875622e696f2f736f6e6172636c6f75642d6769746875622d7374617469632d7265736f75726365732f76322f636f6d6d6f6e2f6661696c65642d313670782e706e67) [1 Security Hotspot](https://sonarcloud.io/project/security_hotspots?id=FreeRTOS_FreeRTOS-Kernel&pullRequest=1203&issueStatuses=OPEN,CONFIRMED&sinceLeakPeriod=true)
> 
> [See analysis details on SonarQube Cloud](https://sonarcloud.io/dashboard?id=FreeRTOS_FreeRTOS-Kernel&pullRequest=1203)

@jasonpcarroll Looks like I would need an additional review on SonarQube for using `strncpy`. Thanks.

sudeep-mohanty commented:
@jasonpcarroll Could you let me know if this PR is reviewable as is or do I need to make some changes for the checks to pass. Thanks.

aggarg commented:
We try to avoid `snprintf` class of functions because it might bloat the code size and use a lot of stack. Only stats formatting functions use `snprintf` which is more of a debugging aid. I'd like to avoid snprintf or string manipulation functions for IDLE task name which is also a debugging aid. What do you think about the following patch - [suggestion.patch](https://github.com/user-attachments/files/18165053/suggestion.patch)?

sudeep-mohanty commented:
Thanks for the inputs @aggarg! I've incorporated your suggestions but since it did not address the out-of-bound mem copy operation warnings flagged by static code analyzers, I have added a fix for it as well. PTAL. Thanks.

sudeep-mohanty commented:
Updated corresponding unit test in [FreeRTOS PR #1314](https://github.com/FreeRTOS/FreeRTOS/pull/1314).

sudeep-mohanty commented:
Thank you for helping merge the PR, @aggarg!
